gateway/src/apicast/policy/rate_limiting_to_service/rate_limiting_to_service.lua

```diff
@@ -0,0 +1,196 @@
+    return resty_limit_req.new(shdict_key, config.rate, config.burst)
+  end,
+  fixed_window = function(config)
+    return resty_limit_count.new(shdict_key, config.count, config.window)
+  end
+}
+
+local function try(f, catch_f)
+  local status, exception = pcall(f)
+  if not status then
+    catch_f(exception)
+  end
+end
+
+local function init_limiter(config)
+  local lim, limerr
+  try(
+    function()
+      lim, limerr = traffic_limiters[config.name](config)
+    end,
+    function(e)
+      return nil, e
+    end
+  )
+  if not lim then
+    return nil, limerr
+  end
+
+  return lim, nil
+end
+
+local function redis_shdict(url)
+  local options = { url = url }
+  local redis, err = ts.connect_redis(options)
+  if not redis then
+    return nil, err
+  end
+
+  return {
+    incr = function(_, key, value, init)
+      if not init then
+        return redis:incrby(key, value), nil
+      end
+      redis:setnx(key, init)
+      return redis:incrby(key, value), nil
+    end,
+    set = function(_, key, value)
+      return redis:set(key, value)
+    end,
+    expire = function(_, key, exptime)
+      local ret = redis:expire(key, exptime)
+      if ret == 0 then
+        return nil, "not found"
+      end
+      return true, nil
+    end,
+    get = function(_, key)
+      local val = redis:get(key)
+      if type(val) == "userdata" then
+        return nil
+      end
+      return val
+    end
+  }
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.limiters = config.limiters
+  self.redis_url = config.redis_url
+
+  return self
+end
+
+function _M:access()
+  local limiters = {}
+  local keys = {}
+
+  if not self.redis_url then
+    -- Only one (the first) limiter is enable.
+    -- Key will be shdict_key ('limiter').
+    local lim, initerr = init_limiter(self.limiters[1])
+    if not lim then
+      ngx.log(ngx.ERR, "unknown limiter: ", initerr)
+      return ngx.exit(500)
+    end
+
+    limiters[1] = lim
+    keys[1] = shdict_key
+
+  else
+    for _, limiter in ipairs(self.limiters) do
+      local lim, initerr = init_limiter(limiter)
+      if not lim then
+        ngx.log(ngx.ERR, "unknown limiter: ", limiter.name, ", err: ", initerr)
+        return ngx.exit(500)
+      end
+
+      local rediserr
+      lim.dict, rediserr = redis_shdict(self.redis_url)
+      if not lim.dict then
+        ngx.log(ngx.ERR, "failed to connect Redis: ", rediserr)
+        return ngx.exit(500)
+      end
+
+      limiters[#limiters + 1] = lim
+      keys[#keys + 1] = limiter.key
+
+    end
+  end
+
+  local states = {}
+  local connections_committed = {}
+  local keys_committed = {}
+
+  local delay, comerr = limit_traffic.combine(limiters, keys, states)
+  if not delay then
+    if comerr == "rejected" then
+      ngx.log(ngx.ERR, "Requests over the limit.")
+      return ngx.exit(429)
+    end
+    ngx.log(ngx.ERR, "failed to limit traffic: ", comerr)
+    return ngx.exit(500)
```

Comment:
I think this might be another configuration option. I'm not sure that everyone wants to respond with an error when there is some issue with rate limiting. Some people might want to let the request go through.